File: open_mahjong_server/server/game_calculation/riichi/riichi_tingpai_check.py
"""
立直麻将听牌检查：一般型（借自古典）+ 七对子 + 国士无双
- 赤 5 (105/205/305) 先归一化为 15/25/35 再进行形状匹配
"""
from typing import List, Set
from time import time


def _normalize_tile(t: int) -> int:
    if t == 105:
        return 15
    if t == 205:
        return 25
    if t == 305:
        return 35
    return t


class PlayerTiles:
    def __init__(self, tiles_list: List[int], combination_list: List[str], complete_step: int):
        self.hand_tiles = sorted(tiles_list)
        self.combination_list = combination_list
        self.complete_step = complete_step

    def __deepcopy__(self, memo):
        return PlayerTiles(self.hand_tiles[:], self.combination_list[:], self.complete_step)


class Riichi_Tingpai_Check:
    yaojiu = {11, 19, 21, 29, 31, 39, 41, 42, 43, 44, 45, 46, 47}

    def __init__(self):
        self.waiting_tiles: Set[int] = set()

    def check_waiting_tiles(self, player_tiles: PlayerTiles):
        self.waiting_tiles.clear()

        if len(player_tiles.hand_tiles) == 13 and not player_tiles.combination_list:
            self.GS_check(player_tiles.hand_tiles)
            self.QD_check(player_tiles.hand_tiles)

        self.normal_check(player_tiles)
        return self.waiting_tiles

    def GS_check(self, hand_tiles: List[int]):
        GS_step_set = set()
        GS_allowed = True
        for tile_id in hand_tiles:
            if tile_id in self.yaojiu:
                GS_step_set.add(tile_id)
            else:
                GS_allowed = False
        if GS_allowed:
            if len(GS_step_set) == 12:
                for i in self.yaojiu:
                    if i not in hand_tiles:
                        self.waiting_tiles.add(i)
            elif len(GS_step_set) == 13:
                for i in self.yaojiu:
                    self.waiting_tiles.add(i)

    def QD_check(self, hand_tiles: List[int]):
        tile_counts = {}
        for tile_id in hand_tiles:
            tile_counts[tile_id] = tile_counts.get(tile_id, 0) + 1

        single = 0
        waiting_tile = None
        for tile_id, count in tile_counts.items():
            if count == 1 or count == 3:
                single += 1
                waiting_tile = tile_id
            if single >= 2:
                return
        if single == 1:
            self.waiting_tiles.add(waiting_tile)

    def normal_check(self, player_tiles: PlayerTiles):
        if not self.normal_check_block(player_tiles):
            return

        all_list = self.normal_check_traverse_quetou(player_tiles)
        end_list = []
        while all_list:
            temp_list = all_list.pop()
            self.normal_check_traverse_kezi(temp_list, all_list)
            self.normal_check_traverse_dazi(temp_list, all_list)
            if temp_list.complete_step >= 11:
                end_list.append(temp_list)

        for i in end_list:
            if len(i.hand_tiles) == 1:
                self.waiting_tiles.add(i.hand_tiles[0])
            elif len(i.hand_tiles) == 2:
                if i.hand_tiles[0] == i.hand_tiles[1]:
                    self.waiting_tiles.add(i.hand_tiles[0])
                elif i.hand_tiles[0] == i.hand_tiles[1] - 1:
                    if i.hand_tiles[0] - 1 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] - 1)
                    if i.hand_tiles[0] + 2 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] + 2)
                elif i.hand_tiles[0] == i.hand_tiles[1] - 2:
                    if i.hand_tiles[0] + 1 < 40:
                        self.waiting_tiles.add(i.hand_tiles[0] + 1)

    def normal_check_block(self, player_tiles: PlayerTiles):
        if not player_tiles.hand_tiles:
            return True
        block_count = len(player_tiles.combination_list)
        tile_id_pointer = player_tiles.hand_tiles[0]
        for tile_id in player_tiles.hand_tiles:
            if tile_id == tile_id_pointer or tile_id == tile_id_pointer + 1:
                pass
            else:
                block_count += 1
            tile_id_pointer = tile_id
        return block_count <= 6

    def normal_check_traverse_quetou(self, player_tiles: PlayerTiles):
        all_list = []
        quetou_id_pointer = 0
        for tile_id in player_tiles.hand_tiles:
            if player_tiles.hand_tiles.count(tile_id) >= 2 and tile_id != quetou_id_pointer:
                temp_list = player_tiles.__deepcopy__(None)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.complete_step += 2
                temp_list.combination_list.append(f"q{tile_id}")
                all_list.append(temp_list)
                quetou_id_pointer = tile_id
        all_list.append(player_tiles.__deepcopy__(None))
        return all_list

    def normal_check_traverse_kezi(self, player_tiles: PlayerTiles, all_list):
        same_tile_id = 0
        for tile_id in player_tiles.hand_tiles:
            if player_tiles.hand_tiles.count(tile_id) >= 3 and tile_id != same_tile_id:
                temp_list = player_tiles.__deepcopy__(None)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.hand_tiles.remove(tile_id)
                temp_list.complete_step += 3
                temp_list.combination_list.append(f"k{tile_id}")
                all_list.append(temp_list)
                same_tile_id = tile_id

    def normal_check_traverse_dazi(self, player_tiles: PlayerTiles, all_list):
        same_tile_id = 0
        for tile_id in player_tiles.hand_tiles:
            if tile_id <= 40:
                if tile_id + 1 in player_tiles.hand_tiles and tile_id + 2 in player_tiles.hand_tiles and tile_id != same_tile_id:
                    temp_list = player_tiles.__deepcopy__(None)
                    temp_list.hand_tiles.remove(tile_id)
                    temp_list.hand_tiles.remove(tile_id + 1)
                    temp_list.hand_tiles.remove(tile_id + 2)
                    temp_list.complete_step += 3
                    temp_list.combination_list.append(f"s{tile_id + 1}")
                    all_list.append(temp_list)
                    same_tile_id = tile_id

    def tingpai_check(self, hand_tile_list: List[int], combination_list: List[str]) -> Set[int]:
        normalized = [_normalize_tile(t) for t in hand_tile_list]
        test_tiles = PlayerTiles(normalized, combination_list.copy(), len(combination_list) * 3)
        self.check_waiting_tiles(test_tiles)
        self.waiting_tiles = {i for i in self.waiting_tiles if i not in {10, 20, 30, 40}}
        return self.waiting_tiles.copy()


if __name__ == "__main__":
    test_hand_tiles = [11, 19, 21, 29, 31, 39, 41, 42, 43, 44, 45, 46, 47]
    checker = Riichi_Tingpai_Check()
    t0 = time()
    result = checker.tingpai_check(test_hand_tiles, [])
    t1 = time()
    print("Riichi Tingpai Check")
    print("手牌:", test_hand_tiles)
    print("听牌:", sorted(result))
    print("耗时:", t1 - t0)

File: open_mahjong_unity/Assets/Scripts/GameScene/Calculation/CalculationScript/Riichi/RiichiCalc/RiichiTingpaiCheck.cs
using System.Collections.Generic;

namespace Riichi {
    /// <summary>
    /// 立直麻将听牌检测：在古典一般型基础上额外支持七对子与十三幺。
    /// 输入支持赤宝牌 105/205/305，内部先归一化为 15/25/35 再检测。
    /// </summary>
    public class RiichiTingpaiCheck {
        private readonly HashSet<int> _waitingTiles = new HashSet<int>();
        private static readonly HashSet<int> InvalidTiles = new HashSet<int> { 10, 20, 30, 40 };
        private static readonly HashSet<int> Yaojiu = new HashSet<int> {
            11, 19, 21, 29, 31, 39, 41, 42, 43, 44, 45, 46, 47
        };

        public HashSet<int> TingpaiCheck(List<int> handTileList, List<string> combinationList) {
            var normalized = NormalizeHand(handTileList);

            _waitingTiles.Clear();

            if (normalized.Count == 13) {
                GsCheck(normalized);
                QdCheckWithoutDragonSevenPairs(normalized);
            }

            int completeStep = combinationList.Count * 3;
            var pt = new RiichiPlayerTiles(normalized, combinationList, completeStep);
            NormalCheck(pt);

            _waitingTiles.RemoveWhere(t => InvalidTiles.Contains(t));
            return new HashSet<int>(_waitingTiles);
        }

        public static HashSet<int> TingpaiCheckStatic(List<int> handTileList, List<string> combinationList, bool debug = false) {
            var checker = new RiichiTingpaiCheck();
            return checker.TingpaiCheck(handTileList, combinationList);
        }

        private static List<int> NormalizeHand(List<int> handTileList) {
            var res = new List<int>(handTileList.Count);
            foreach (int t in handTileList) {
                if (t == 105) res.Add(15);
                else if (t == 205) res.Add(25);
                else if (t == 305) res.Add(35);
                else res.Add(t);
            }
            res.Sort();
            return res;
        }

        private void GsCheck(List<int> handTiles) {
            var yaoSet = new HashSet<int>();
            foreach (int t in handTiles) {
                if (!Yaojiu.Contains(t)) return;
                yaoSet.Add(t);
            }
            if (yaoSet.Count == 12) {
                foreach (int i in Yaojiu) {
                    if (!handTiles.Contains(i)) _waitingTiles.Add(i);
                }
            } else if (yaoSet.Count == 13) {
                foreach (int i in Yaojiu) _waitingTiles.Add(i);
            }
        }

        /// <summary>
        /// 七对子听牌（不含龙七对 dragon seven pairs）：仅 6 对 + 1 单张。
        /// 5 对 + 1 刻听第 4 张和牌后为同牌四张，非日麻七对子。
        /// </summary>
        private void QdCheckWithoutDragonSevenPairs(List<int> handTiles) {
            var counts = new Dictionary<int, int>();
            foreach (int t in handTiles) {
                counts[t] = counts.TryGetValue(t, out int c) ? c + 1 : 1;
            }
            int single = 0;
            int waitingTile = -1;
            foreach (var kv in counts) {
                if (kv.Value == 1) {
                    single++;
                    waitingTile = kv.Key;
                } else if (kv.Value == 2) {
                    continue;
                } else {
                    return;
                }
                if (single >= 2) {
                    return;
                }
            }
            if (single == 1 && waitingTile >= 0) {
                _waitingTiles.Add(waitingTile);
            }
        }

        private void NormalCheck(RiichiPlayerTiles pt) {
            if (!BlockCheck(pt)) return;
            var allList = TraverseQuetou(pt);
            var endList = new List<RiichiPlayerTiles>();
            while (allList.Count > 0) {
                var cur = allList[allList.Count - 1];
                allList.RemoveAt(allList.Count - 1);
                TraverseKezi(cur, allList);
                TraverseDazi(cur, allList);
                if (cur.CompleteStep >= 11) endList.Add(cur);
            }

            foreach (var item in endList) {
                if (item.HandTiles.Count == 1) {
                    _waitingTiles.Add(item.HandTiles[0]);
                } else if (item.HandTiles.Count == 2) {
                    int a = item.HandTiles[0], b = item.HandTiles[1];
                    if (a == b) {
                        _waitingTiles.Add(a);
                    } else if (a == b - 1) {
                        if (a - 1 > 0 && a - 1 < 40) _waitingTiles.Add(a - 1);
                        if (a + 2 > 0 && a + 2 < 40) _waitingTiles.Add(a + 2);
                    } else if (a == b - 2) {
                        if (a + 1 > 0 && a + 1 < 40) _waitingTiles.Add(a + 1);
                    }
                }
            }
        }

        private static bool BlockCheck(RiichiPlayerTiles pt) {
            if (pt.HandTiles.Count == 0) return true;
            int blockCount = pt.CombinationList.Count;
            int pointer = pt.HandTiles[0];
            foreach (int t in pt.HandTiles) {
                if (t != pointer && t != pointer + 1) blockCount++;
                pointer = t;
            }
            return blockCount <= 6;
        }

        private static List<RiichiPlayerTiles> TraverseQuetou(RiichiPlayerTiles pt) {
            var result = new List<RiichiPlayerTiles>();
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (CountInList(pt.HandTiles, t) >= 2 && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.CompleteStep += 2;
                    copy.CombinationList.Add($"q{t}");
                    result.Add(copy);
                    prevId = t;
                }
            }
            result.Add(pt.DeepCopy());
            return result;
        }

        private static void TraverseKezi(RiichiPlayerTiles pt, List<RiichiPlayerTiles> allList) {
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (CountInList(pt.HandTiles, t) >= 3 && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t);
                    copy.CompleteStep += 3;
                    copy.CombinationList.Add($"k{t}");
                    allList.Add(copy);
                    prevId = t;
                }
            }
        }

        private static void TraverseDazi(RiichiPlayerTiles pt, List<RiichiPlayerTiles> allList) {
            int prevId = 0;
            foreach (int t in pt.HandTiles) {
                if (t <= 40 && pt.HandTiles.Contains(t + 1) && pt.HandTiles.Contains(t + 2) && t != prevId) {
                    var copy = pt.DeepCopy();
                    copy.HandTiles.Remove(t);
                    copy.HandTiles.Remove(t + 1);
                    copy.HandTiles.Remove(t + 2);
                    copy.CompleteStep += 3;
                    copy.CombinationList.Add($"s{t + 1}");
                    allList.Add(copy);
                    prevId = t;
                }
            }
        }

        private static int CountInList(List<int> list, int value) {
            int count = 0;
            for (int i = 0; i < list.Count; i++) {
                if (list[i] == value) count++;
            }
            return count;
        }
    }
}
